Agents/HMLandRegistryAgent/agent/kgutils/stackclients.py

```diff
@@ -0,0 +1,288 @@
+# Date: 14 Feb 2023                            #
+################################################
+
+# The purpose of this module is to provide functionality to interact with the
+# overarching Docker Stack using the Stack Clients wrapped by py4jps
+
+import glob
+import jaydebeapi
+import json
+
+from agent.errorhandling.exceptions import StackException
+from agent.kgutils.javagateway import stackClientsGw, jpsBaseLibGW
+from agent.utils.env_configs import DATABASE, LAYERNAME, GEOSERVER_WORKSPACE, \
+                                    BUILDINGS_TABLE
+from agent.utils.stack_configs import DB_URL, DB_USER, DB_PASSWORD, QUERY_ENDPOINT
+
+# Initialise logger
+from py4jps import agentlogging
+logger = agentlogging.get_logger("prod")
+
+
+class StackClient:
+    # Define parent class for all Stack Clients to minimise number of required
+    # JAVA resource views and import the required java classes
+
+    # Create ONE JPS_BASE_LIB view
+    jpsBaseLib_view = jpsBaseLibGW.createModuleView()
+    jpsBaseLibGW.importPackages(jpsBaseLib_view,"uk.ac.cam.cares.jps.base.query.*")
+
+    # Create ONE Stack Clients view
+    stackClients_view = stackClientsGw.createModuleView()
+    stackClientsGw.importPackages(stackClients_view, "com.cmclinnovations.stack.clients.postgis.PostGISClient")
+    stackClientsGw.importPackages(stackClients_view, "com.cmclinnovations.stack.clients.geoserver.GeoServerClient")
+    stackClientsGw.importPackages(stackClients_view, "com.cmclinnovations.stack.clients.geoserver.GeoServerVectorSettings")
+    stackClientsGw.importPackages(stackClients_view, "com.cmclinnovations.stack.clients.geoserver.UpdatedGSVirtualTableEncoder")
+    # Include required third-party packages
+    stackClientsGw.importPackages(stackClients_view, "it.geosolutions.geoserver.rest.encoder.metadata.virtualtable.VTGeometryEncoder")
+
+
+class PostGISClient(StackClient):
+    
+    def __init__(self, dburl=DB_URL, dbuser=DB_USER, dbpassword=DB_PASSWORD,
+                 database=DATABASE):
+        self.dburl = dburl
+        self.dbuser = dbuser
+        self.dbpassword = dbpassword
+        self.database = database
+        try:
+            # Retrieve JDBC connection properties
+            self.conn_props = self.connection_properties()
+        except Exception as ex:
+            logger.error("Unable to retrieve JDBC connection properties.")
+            raise StackException("Unable to retrieve JDBC connection properties.") from ex
+
+    
+    def connection_properties(self):
+        """
+        This function returns dictionary of JDBC connection properties 
+        to PostgreSQL database
+        NOTE: This function is necessary as stack clients' PostGISClient does not provide
+              any general functionality to execute SQL queries; hence, they need to be
+              defined here and executed using jaydebeapi 
+              (which requires all JDBC properties incl. driver)
+        """
```

Comment:
Not 100% sure if this is relevant but the [`PostGISClient::getRemoteStoreClient`](https://github.com/cambridge-cares/TheWorldAvatar/blob/582-add-dcat-concepts-in-data-uploader/Deploy/stacks/dynamic/stack-clients/src/main/java/com/cmclinnovations/stack/clients/postgis/PostGISClient.java#L69) methods will return a preconfigured `RemoteRDBStoreClient` object that should be able to make SQL requests.

Reply:
I am aware of this (and we should look into this in more detail in the future). Last time I tried it (which is a few months ago though), I did not get it working and still needed the JDBC driver and connection properties to get it working. I suggest, this shall be addressed in a future iteration of the agent